### PR TITLE
fix: Fixed a bug where double ? would be generated for stubs

### LINKED ISSUE #87: Double `?` in generated types
### Describe the bug

The callback [`callbackOnBatchCompletion`](https://library.safeds.com/en/latest/reference/safeds/ml/nn/ClassificationNeuralNetwork/#safeds.ml.nn.ClassificationNeuralNetwork.fit) has the type 

```txt
Callable[[int, float], None] | None
```

This is currently translated to

```txt
(param1: Int, param2: Float) -> param3: Nothing??
```

### To Reproduce

/

### Expected behavior

The expected output would be 

```txt
union<(param1: Int, param2: Float) -> param3: Nothing?, Nothing?>
```

Only named types (class/enum/enum variant) support the `?` syntax for nullability.

### Screenshots (optional)

_No response_

### Additional Context (optional)

_No response_

### PR DESCRIPTION
**Pull request by Masara**
Closes #87

### Summary of Changes

Fixed the bug described in #87.

### CODE REVIEW

**Review by lars-reimann — CHANGES_REQUESTED** [review]
This does not fully fix the issue. `Literal["Test"] | None` gets translated to `literal<"Test">?` for example. The `?` syntax for nullability is only supported for named Safe-DS types (class types/enum types/enum variant types/type parameter types). Out of those, you probably only generate class types.

For this literal type you could create `literal<"Test", null>` instead. For everything else, you need the `union` with `Nothing?`.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[type_var_module].sdsstub [hunk 1]
```diff
@@ -75,13 +75,13 @@ class CollectionTypeVar2<T>() {
 }
 
 class MappingTypeVar(
-    data: Map<String, Any>? = null
+    data: union<Map<String, Any>, Nothing?> = null
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[type_var_module].sdsstub:78]: This was fine before.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[type_var_module].sdsstub [hunk 2]
```diff
@@ -75,13 +75,13 @@ class CollectionTypeVar2<T>() {
...
 )
 
 class MappingTypeVar2(
     data: Map<String, List<Any>>
 )
 
 class MappingTypeVar3<T>(
-    data: Map<String, List<T>>?
+    data: union<Map<String, List<T>>, Nothing?>
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[type_var_module].sdsstub:86]: This was fine before.

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub [hunk 3]
```diff
@@ -30,7 +30,14 @@ fun params(
     optional: Int?,
     @PythonName("tuple_") tuple: Tuple<Int, String, Boolean>,
     `literal`: literal<"Some String">,
-    @PythonName("any_") any: Any
+    @PythonName("any_") any: Any,
+    @PythonName("callable_none") callableNone: union<(param1: Int, param2: Float) -> result1: Nothing?, Nothing?>,
+    @PythonName("literal_none") literalNone: union<Nothing?, literal<"1", 2>>,
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub:35]:
```suggestion
    @PythonName("literal_none") literalNone: literal<"1", 2, null>,
```

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub [hunk 4]
```diff
@@ -30,7 +30,14 @@ fun params(
...
+    @PythonName("set_none") setNone: union<Nothing?, Set<Int>>,
+    @PythonName("dict_none") dictNone: union<Map<String, Int>, Nothing?>,
+    @PythonName("named_class_none") namedClassNone: FunctionModuleClassA?,
+    @PythonName("list_class_none") listClassNone: union<List<Float>, Nothing?>,
+    @PythonName("tuple_class_none") tupleClassNone: union<Nothing?, Tuple<Int, String>>
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub:40]: Would it be possible to consistently add `Nothing?` at the end of the union?

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub [hunk 5]
```diff
@@ -30,7 +30,14 @@ fun params(
...
     @PythonName("tuple_") tuple: Tuple<Int, String, Boolean>,
     `literal`: literal<"Some String">,
-    @PythonName("any_") any: Any
+    @PythonName("any_") any: Any,
+    @PythonName("callable_none") callableNone: union<(param1: Int, param2: Float) -> result1: Nothing?, Nothing?>,
+    @PythonName("literal_none") literalNone: union<Nothing?, literal<"1", 2>>,
+    @PythonName("set_none") setNone: union<Nothing?, Set<Int>>,
+    @PythonName("dict_none") dictNone: union<Map<String, Int>, Nothing?>,
+    @PythonName("named_class_none") namedClassNone: FunctionModuleClassA?,
+    @PythonName("list_class_none") listClassNone: union<List<Float>, Nothing?>,
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub:39]:
```suggestion
    @PythonName("list_class_none") listClassNone: List<Float>?,
```

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub [hunk 6]
```diff
@@ -30,7 +30,14 @@ fun params(
     optional: Int?,
     @PythonName("tuple_") tuple: Tuple<Int, String, Boolean>,
     `literal`: literal<"Some String">,
-    @PythonName("any_") any: Any
+    @PythonName("any_") any: Any,
+    @PythonName("callable_none") callableNone: union<(param1: Int, param2: Float) -> result1: Nothing?, Nothing?>,
+    @PythonName("literal_none") literalNone: union<Nothing?, literal<"1", 2>>,
+    @PythonName("set_none") setNone: union<Nothing?, Set<Int>>,
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub:36]:
```suggestion
    @PythonName("set_none") setNone: Set<Int>?,
```

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub [hunk 7]
```diff
@@ -30,7 +30,14 @@ fun params(
...
+    @PythonName("dict_none") dictNone: union<Map<String, Int>, Nothing?>,
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub:37]:
```suggestion
    @PythonName("dict_none") dictNone: Map<String, Int>?,
```

File: tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub [hunk 8]
```diff
@@ -83,7 +90,7 @@ fun specialParams(
     @PythonName("none_bool_none_union") noneBoolNoneUnion: Boolean?,
     @PythonName("none_bool_int_union") noneBoolIntUnion: union<Boolean, Int, Nothing?>,
     @PythonName("none_none_bool_none_union") noneNoneBoolNoneUnion: Boolean?,
-    @PythonName("none_list_union_none_none") noneListUnionNoneNone: List<Nothing?>?,
+    @PythonName("none_list_union_none_none") noneListUnionNoneNone: union<List<Nothing?>, Nothing?>,
```
Comment (lars-reimann) [inline tests/safeds_stubgen/stubs_generator/__snapshots__/test_generate_stubs/TestStubFileGeneration.test_stub_creation[function_module].sdsstub:93]: This was fine before.

**Review by lars-reimann — CHANGES_REQUESTED** [review]

**Comment by Masara** [comment]
I hope everything is all right now

**Comment by lars-reimann** [comment]
Looks good now, I'll review it tomorrow evening if the conflicts are resolved.

**Review by lars-reimann — APPROVED** [review]
👍

**Pull request merged**